### PR TITLE
Issue/345/default cosmological model

### PR DESCRIPTION
**Pull request by pennalima**

### CODE REVIEW

**Comment by combet** [comment]
Thanks @pennalima, @vitenti!

I saw you made `ccl` the default backend in that branch, which I think is a good idea. However, the documentation is not building anymore and this is related to what @hsinfan1996 noticed in #365, and the ongoing discussion with @m-aguena there.

**Comment by m-aguena** [comment]
@combet I added @vitenti 's suggestion (https://github.com/LSSTDESC/CLMM/pull/365#issuecomment-769710572) and it did not seem to fix it. Looking at the errors, it does not look to be the source of the error.

**Comment by hsinfan1996** [comment]
> @combet I added @vitenti 's suggestion ([#365 (comment)](https://github.com/LSSTDESC/CLMM/pull/365#issuecomment-769710572)) and it did not seem to fix it. Looking at the errors, it does not look to be the source of the error.

In clmm/theory/ccl.py line 51, change CCLCosmology to ccl.Cosmology and the documentation builds, but I don't know if it is the right way to fix it. I just did it for some test.

**Comment by m-aguena** [comment]
I found one solution, but I am not sure if it is the correct one:
https://github.com/LSSTDESC/CLMM/pull/365#issuecomment-769840824

File: clmm/__init__.py [hunk 1]
```diff
@@ -6,5 +6,10 @@
 from .theory import compute_reduced_shear_from_convergence, compute_3d_density, compute_surface_density, compute_excess_surface_density, compute_critical_surface_density, compute_tangential_shear, compute_convergence, compute_reduced_tangential_shear, Modeling, Cosmology
 from . import support
 
+import os
```
Comment (m-aguena) [inline clmm/__init__.py:9]: did you mean to add these changes?

**Review by m-aguena — COMMENTED** [review]
Overall, everything is good. Just some minor comments

**Comment by hsinfan1996** [comment]
Are we allowed to build documentation with only one of the backends installed?

**Comment by combet** [comment]
@pennalima @m-aguena - when running pytest locally, some tests are failing as the required tolerance are not quite enough. And I don't have the same tests failing whether I run on my laptop or at CC-in2p3. But GA does not seem to complain, so I wonder if the precision reached can depend on the system on which the tests are run? Anyone noticed anything like that?

**Comment by combet** [comment]
> Given the changes in this branch, are we allowed to build documentation with only one of the backends installed?

From what I've seen, the changes allow to solve the errors from CCL or NC when building the documentation, but it still needs cluster-toolkit to be installed.

**Comment by m-aguena** [comment]
@combet this is strange, I did not find any error on my laptop. But on nersc, I got an even more basic error:
```
modeling_data = {'cosmo_reltol': 1e-05, 'dataops_reltol': 5e-06, 'nick': 'ct', 'theory_reltol': 0.0035}, cosmo_init = {'H0': 67.0, 'Omega_b0': 0.06, 'Omega_dm0': 0.22, 'Omega_k0': 0.01}

    def test_cosmo_basic(modeling_data, cosmo_init):
        """ Unit tests abstract class z and a methdods """
        reltol = modeling_data['cosmo_reltol']
        cosmo = theo.Cosmology(**cosmo_init)
        # Test get_<PAR>(z)
        Omega_m0 = cosmo['Omega_m0']
        assert_allclose(cosmo.get_Omega_m(0.0), Omega_m0, **TOLERANCE)
        assert_allclose(cosmo.get_E2Omega_m(0.0), Omega_m0, **TOLERANCE)
        # Test getting all parameters
        for param in ("Omega_m0", "Omega_b0", "Omega_dm0", "Omega_k0", 'h', 'H0'):
            cosmo[param]
        # Test params values
        for param in cosmo_init.keys():
>           assert_allclose(cosmo_init[param], cosmo[param], **TOLERANCE)
E           AssertionError: 
E           Not equal to tolerance rtol=1e-15, atol=0
E           
E           Mismatched elements: 1 / 1 (100%)
E           Max absolute difference: 1.11022302e-16
E           Max relative difference: 1.11022302e-14
E            x: array(0.01)
E            y: array(0.01)

tests/test_cosmo_parent.py:111: AssertionError
```
This could be a python version difference (`3.9.0` on my laptop and `3.8.5` on nersc)

**Comment by combet** [comment]
> @combet this is strange, I did not find any error on my laptop. But on nersc, I got an even more basic error:
> 
> ```
> modeling_data = {'cosmo_reltol': 1e-05, 'dataops_reltol': 5e-06, 'nick': 'ct', 'theory_reltol': 0.0035}, cosmo_init = {'H0': 67.0, 'Omega_b0': 0.06, 'Omega_dm0': 0.22, 'Omega_k0': 0.01}
> 
>     def test_cosmo_basic(modeling_data, cosmo_init):
>         """ Unit tests abstract class z and a methdods """
>         reltol = modeling_data['cosmo_reltol']
>         cosmo = theo.Cosmology(**cosmo_init)
>         # Test get_<PAR>(z)
>         Omega_m0 = cosmo['Omega_m0']
>         assert_allclose(cosmo.get_Omega_m(0.0), Omega_m0, **TOLERANCE)
>         assert_allclose(cosmo.get_E2Omega_m(0.0), Omega_m0, **TOLERANCE)
>         # Test getting all parameters
>         for param in ("Omega_m0", "Omega_b0", "Omega_dm0", "Omega_k0", 'h', 'H0'):
>             cosmo[param]
>         # Test params values
>         for param in cosmo_init.keys():
> >           assert_allclose(cosmo_init[param], cosmo[param], **TOLERANCE)
> E           AssertionError: 
> E           Not equal to tolerance rtol=1e-15, atol=0
> E           
> E           Mismatched elements: 1 / 1 (100%)
> E           Max absolute difference: 1.11022302e-16
> E           Max relative difference: 1.11022302e-14
> E            x: array(0.01)
> E            y: array(0.01)
> 
> tests/test_cosmo_parent.py:111: AssertionError
> ```
> 
> This could be a python version difference (`3.9.0` on my laptop and `3.8.5` on nersc)

Oh, yes, this is the one I get running at cc-in2p3, where I run python 3.7

**Comment by combet** [comment]
On my laptop, running 3.8.6, I also get:
```
modeling_data = {'cosmo_reltol': 1e-08, 'dataops_reltol': 1e-08, 'nick': 'nc', 'theory_reltol': 1e-08}

    def test_instantiate(modeling_data):
        """ Unit tests for modeling objects' instantiation """
    
        m = theo.Modeling()
        m.set_concentration(4.0)
        m.set_mass(1.0e15)
        assert m.backend == theo.be_nick
    
        assert_raises(ValueError, m.set_cosmo, 3.0)
        assert_raises(ValueError, m.set_halo_density_profile, halo_profile_model='bla')
        assert_raises(ValueError, m.set_halo_density_profile, massdef='blu')
    
        if theo.be_nick == 'nc':
            import gi
            gi.require_version("NumCosmoMath", "1.0")
            gi.require_version("NumCosmo", "1.0")
    
            import gi.repository.NumCosmoMath as Ncm
            import gi.repository.NumCosmo as Nc
    
            mset = m.get_mset()
            assert isinstance(mset, Ncm.MSet)
    
            m.set_mset(mset)
            assert_raises(AttributeError, m.set_mset, 3)
    
        r_proj = np.logspace(-2, 2, 100)
        z_cl = 0.5
        z_src = 0.85
        Sigma = m.eval_surface_density(r_proj, z_cl)
        Sigma_mean = m.eval_mean_surface_density(r_proj, z_cl)
        Sigma_excess = m.eval_excess_surface_density(r_proj, z_cl)
    
        assert_allclose(Sigma_excess, (Sigma_mean-Sigma), rtol=5.0e-15)
    
        shear = m.eval_tangential_shear(r_proj, z_cl, z_src)
        convergence = m.eval_convergence(r_proj, z_cl, z_src)
        reduced_shear = m.eval_reduced_tangential_shear(r_proj, z_cl, z_src)
        magnification = m.eval_magnification(r_proj, z_cl, z_src)
    
>       assert_allclose(reduced_shear, shear/(1.0-convergence), rtol=8.0e-15)
E       AssertionError: 
E       Not equal to tolerance rtol=8e-15, atol=0
E       
E       Mismatched elements: 4 / 100 (4%)
E       Max absolute difference: 3.47499807e-14
E       Max relative difference: 2.0581259e-13
E        x: array([3.755258e-01, 3.509255e-01, 3.293310e-01, 3.102190e-01,
E              2.931805e-01, 2.778907e-01, 2.640886e-01, 2.515618e-01,
E              2.401355e-01, 2.296645e-01, 2.200270e-01, 2.111199e-01,...
E        y: array([3.755258e-01, 3.509255e-01, 3.293310e-01, 3.102190e-01,
E              2.931805e-01, 2.778907e-01, 2.640886e-01, 2.515618e-01,
E              2.401355e-01, 2.296645e-01, 2.200270e-01, 2.111199e-01,...

tests/test_theory_parent.py:70: AssertionError
```

**Comment by vitenti** [comment]
> On my laptop, running 3.8.6, I also get:
> 
> ```
> modeling_data = {'cosmo_reltol': 1e-08, 'dataops_reltol': 1e-08, 'nick': 'nc', 'theory_reltol': 1e-08}
> 
>     def test_instantiate(modeling_data):
>         """ Unit tests for modeling objects' instantiation """
>     
>         m = theo.Modeling()
>         m.set_concentration(4.0)
>         m.set_mass(1.0e15)
>         assert m.backend == theo.be_nick
>     
>         assert_raises(ValueError, m.set_cosmo, 3.0)
>         assert_raises(ValueError, m.set_halo_density_profile, halo_profile_model='bla')
>         assert_raises(ValueError, m.set_halo_density_profile, massdef='blu')
>     
>         if theo.be_nick == 'nc':
>             import gi
>             gi.require_version("NumCosmoMath", "1.0")
>             gi.require_version("NumCosmo", "1.0")
>     
>             import gi.repository.NumCosmoMath as Ncm
>             import gi.repository.NumCosmo as Nc
>     
>             mset = m.get_mset()
>             assert isinstance(mset, Ncm.MSet)
>     
>             m.set_mset(mset)
>             assert_raises(AttributeError, m.set_mset, 3)
>     
>         r_proj = np.logspace(-2, 2, 100)
>         z_cl = 0.5
>         z_src = 0.85
>         Sigma = m.eval_surface_density(r_proj, z_cl)
>         Sigma_mean = m.eval_mean_surface_density(r_proj, z_cl)
>         Sigma_excess = m.eval_excess_surface_density(r_proj, z_cl)
>     
>         assert_allclose(Sigma_excess, (Sigma_mean-Sigma), rtol=5.0e-15)
>     
>         shear = m.eval_tangential_shear(r_proj, z_cl, z_src)
>         convergence = m.eval_convergence(r_proj, z_cl, z_src)
>         reduced_shear = m.eval_reduced_tangential_shear(r_proj, z_cl, z_src)
>         magnification = m.eval_magnification(r_proj, z_cl, z_src)
>     
> >       assert_allclose(reduced_shear, shear/(1.0-convergence), rtol=8.0e-15)
> E       AssertionError: 
> E       Not equal to tolerance rtol=8e-15, atol=0
> E       
> E       Mismatched elements: 4 / 100 (4%)
> E       Max absolute difference: 3.47499807e-14
> E       Max relative difference: 2.0581259e-13
> E        x: array([3.755258e-01, 3.509255e-01, 3.293310e-01, 3.102190e-01,
> E              2.931805e-01, 2.778907e-01, 2.640886e-01, 2.515618e-01,
> E              2.401355e-01, 2.296645e-01, 2.200270e-01, 2.111199e-01,...
> E        y: array([3.755258e-01, 3.509255e-01, 3.293310e-01, 3.102190e-01,
> E              2.931805e-01, 2.778907e-01, 2.640886e-01, 2.515618e-01,
> E              2.401355e-01, 2.296645e-01, 2.200270e-01, 2.111199e-01,...
> 
> tests/test_theory_parent.py:70: AssertionError
> ```

What happens is that rounding errors are machine, compiler and optimization dependent. So, in these cases (errors close to the machine epsilon = 2.2204460492503131e-16 (double)) we should require less stringent limits.

**Comment by combet** [comment]
Thanks @vitenti for the explanation - so I relaxed the requirement from 1e-15 to 1e-12, which I believe is still really sufficient for what we're looking for. @m-aguena, could you confirm this solved the problem at NERSC as well. With this, things are OK on both my laptop and my setup at CC-in2p3.

**Comment by m-aguena** [comment]
@combet yes, it ran ok on NERSC now

**Comment by hsinfan1996** [comment]
> > Given the changes in this branch, are we allowed to build documentation with only one of the backends installed?
> 
> @hsinfan1996 From what I've seen, the changes have solved the errors from CCL or NC when building the documentation, this fixing the GA issue, but cluster-toolkit still needs to be installed for the documentation build process to reach the end.

I see.

**Comment by combet** [comment]
@pennalima, pinging you in case these fell into the cracks :) Could you please address the 3 questions from @m-aguena above (starting [here](https://github.com/LSSTDESC/CLMM/pull/371#discussion_r566996018)), so we may move forward with this PR.

File: clmm/cosmology/numcosmo.py [hunk 2]
```diff
@@ -33,23 +33,32 @@ def __init__(self, dist=None, dist_zmax=15.0, **kwargs):
...
+        assert isinstance(be_cosmo, Nc.HICosmoDECpl)
         assert isinstance(be_cosmo.peek_reparam (), Nc.HICosmoDEReparamOk)
         self.be_cosmo = be_cosmo
 
     def _init_from_params(self, H0, Omega_b0, Omega_dm0, Omega_k0):
 
-        self.be_cosmo = Nc.HICosmo.new_from_name(Nc.HICosmo, "NcHICosmoDEXcdm")
-        self.be_cosmo.param_set_lower_bound(Nc.HICosmoDESParams.T_GAMMA0, 0.0)
+        self.be_cosmo = Nc.HICosmo.new_from_name(Nc.HICosmo, "NcHICosmoDECpl{'massnu-length':<1>}")
+        #self.be_cosmo.param_set_lower_bound(Nc.HICosmoDESParams.T_GAMMA0, 0.0)
```
[the remark this answers is not in the document]
Comment (vitenti) [reply clmm/cosmology/numcosmo.py:43]: No that was a compatibility line to emulate models without radiation.

File: clmm/__init__.py (hunk 1 above)
Comment (vitenti) [reply clmm/__init__.py:9]: These are leftovers from testing phase.

File: tests/test_mockdata.py [hunk 3]
```diff
@@ -9,6 +9,7 @@
 from clmm.support.sampler import fitters
 from clmm import Cosmology
 
+import os
```
[the remark this answers is not in the document]
Comment (vitenti) [reply tests/test_mockdata.py:12]: We'll do.

**Comment by vitenti** [comment]
Hello @combet , I fixed all points raised by @m-aguena.

**Review by combet — APPROVED** [review]
Thank you @pennalima, @vitenti - all looks good.

**Pull request merged**